lib/dao.js

```diff
@@ -3089,7 +3089,7 @@ DataAccessObject.replaceById = function(id, data, options, cb) {
 
         function replaceCallback(err, data) {
           if (err) return cb(err);
-          context.data = data;
+          if (data) context.data = data;
```

Comment:
Is it better to move this line after calling `generateContextData`, e.g. 

```
if (typeof connector.generateContextData === 'function') {
   context = connector.generateContextData(context, data);	                       
} else {
  if (data) context.data = data;
}
```

In this way, `generateContextData(context, data)` will continue to get old data in `context`. Otherwise, I think implementation of `generateContextData` is likely to be broken.

Reply:
I think you are right, @shenghu.

In the original pull request #1587 contributed by @juliendossantos, the patch looked like this:

```diff
           if (typeof connector.generateContextData === 'function') {
             context = connector.generateContextData(context, data);
+          } else {
+            context.data = data;
           }
```

I changed it to what we have now, because it seemed to be a simpler solution and frankly I did not really understand the difference and possible ramifications.